esphome/components/ili9341/ili9341_display.cpp

```diff
@@ -0,0 +1,239 @@
+  this->set_addr_window_(0, 0, this->get_width_internal(), this->get_height_internal());
+  this->start_data_();
+
+  auto color565 = color.to_rgb_565();
+  for (uint32_t i = 0; i < (this->get_width_internal()) * (this->get_height_internal()); i++) {
+    this->write_byte(color565 >> 8);
+    this->write_byte(color565);
+    buffer_[i] = 0;
+  }
+  this->end_data_();
+}
+
+void HOT ILI9341Display::draw_absolute_pixel_internal(int x, int y, Color color) {
+  if (x >= this->get_width_internal() || x < 0 || y >= this->get_height_internal() || y < 0)
+    return;
+
+  // low and high watermark may speed up drawing from buffer
+  this->x_low_ = (x < this->x_low_) ? x : this->x_low_;
+  this->y_low_ = (y < this->y_low_) ? y : this->y_low_;
+  this->x_high_ = (x > this->x_high_) ? x : this->x_high_;
+  this->y_high_ = (y > this->y_high_) ? y : this->y_high_;
+
+  uint32_t pos = (y * width_) + x;
+  auto color565 = color.to_rgb_565();
+  buffer_[pos] = convert_to_8bit_color_(color565);
+}
+
+// should return the total size: return this->get_width_internal() * this->get_height_internal() * 2 // 16bit color
+// values per bit is huge
+uint32_t ILI9341Display::get_buffer_length_() { return this->get_width_internal() * this->get_height_internal(); }
+
+void ILI9341Display::start_command_() {
+  this->dc_pin_->digital_write(false);
+  this->enable();
+}
+
+void ILI9341Display::end_command_() { this->disable(); }
+void ILI9341Display::start_data_() {
+  this->dc_pin_->digital_write(true);
+  this->enable();
+}
+void ILI9341Display::end_data_() { this->disable(); }
+
+void ILI9341Display::init_lcd_(const uint8_t *init_cmd) {
+  uint8_t cmd, x, num_args;
+  const uint8_t *addr = init_cmd;
+  while ((cmd = pgm_read_byte(addr++)) > 0) {
+    x = pgm_read_byte(addr++);
+    num_args = x & 0x7F;
+    send_command(cmd, addr, num_args);
+    addr += num_args;
+    if (x & 0x80)
+      delay(150);  // NOLINT
+  }
+}
+
+void ILI9341Display::set_addr_window_(uint16_t x1, uint16_t y1, uint16_t w, uint16_t h) {
+  uint16_t x2 = (x1 + w - 1), y2 = (y1 + h - 1);
+  this->command(ILI9341_CASET);  // Column address set
+  this->start_data_();
+  this->write_byte(x1 >> 8);
+  this->write_byte(x1);
+  this->write_byte(x2 >> 8);
+  this->write_byte(x2);
+  this->end_data_();
+  this->command(ILI9341_PASET);  // Row address set
+  this->start_data_();
+  this->write_byte(y1 >> 8);
+  this->write_byte(y1);
+  this->write_byte(y2 >> 8);
+  this->write_byte(y2);
+  this->end_data_();
+  this->command(ILI9341_RAMWR);  // Write to RAM
+}
+
+void ILI9341Display::invert_display_(bool invert) { this->command(invert ? ILI9341_INVON : ILI9341_INVOFF); }
+
+int ILI9341Display::get_width_internal() { return this->width_; }
+int ILI9341Display::get_height_internal() { return this->height_; }
+
+//   M5Stack display
+void ILI9341M5Stack::initialize() {
+  this->init_lcd_(INITCMD_M5STACK);
+  this->width_ = 320;
+  this->height_ = 240;
+  this->fill_internal_(COLOR_BLACK);
```

Comment:
You need to invert M5Stack display:

```suggestion
  this->invert_display_(true);
  this->fill_internal_(COLOR_BLACK);
```

Source: compiled your source branch and flashed it on M5Stack Grey.

|Without invert|With invert|
|---|---|
|![IMG_0410](https://user-images.githubusercontent.com/1821846/89304074-72e82680-d686-11ea-90a2-783cd850d21c.jpg)|![IMG_0411](https://user-images.githubusercontent.com/1821846/89304138-885d5080-d686-11ea-87b7-dec7712b0472.jpg)|

Reply:
Thanks for your suggestion, patch applied! I don't have an M5 stack module so there is no way for me to check it, but it's nice to know that this code works for you as well 👍